### ISSUE TITLE
Playback speed doesn't change

### ISSUE THREAD

**Issue by DC-Modding** [jellyfin/jellyfin-androidtv#3881]
### Describe the bug

1. Start playing any media
2. Click on "playback speed" setting
3. Select any setting (0.25x-2.00x)

The playback speed doesn't change. Doesn't matter if you select slower or faster

### Application version

0.17.2

### Device information

Zidoo Z9X pro

### Android version

Android 11

### Jellyfin server version

10.9.9

**Comment by FunctionDJ** [comment]
I feel like it's a recent regression. I almost always watch at 1.25x and only noticed today that the setting stopped working. It was working a couple of days ago. Not sure when my app updated (Play Store).

**Comment by DC-Modding** [comment]
Think it broke with the 0.17 update from 0.16.x
At least I noticed it right after it.

**Comment by Kuphi** [comment]
Duplicate of https://github.com/jellyfin/jellyfin-androidtv/issues/2367
Maybe reopen https://github.com/jellyfin/jellyfin-androidtv/issues/2367, because it has not been resolved so far.

**Comment by Kuphi** [comment]
I was able to change the playback speed by building jellyfin-androidtv from master branch. This gave me a new option in the settings in the jellyfin-androidtv app:
Settings -> Developer Settings -> "Prefer FFmpeg for audio playback" (check the box)

But due to circumstances of Amazon Fire TV Sticks, this led to another problem:
>"Interestingly, once the frames are "late" (=the presentation time is in the past), the decoder is suddenly able to quickly release all decoded frames, so that no frames are actually dropped, but you get this jerky, jumpy movements in regular intervals."

@tonihei posted a solution two years ago, at least for Amazon Fire TV Sticks:
https://github.com/google/ExoPlayer/issues/9773#issuecomment-1004683089

If anyone can help to transfer this solution to jellyfin-androidtv that would be great.

**Comment by nielsvanvelzen** [comment]
Don't bother upstream developers. This is not appreciated.

**Comment by Kuphi** [comment]
> Don't bother upstream developers. This is not appreciated.

I'm genuinely trying to contribute to solving this issue, but I find your comment a bit difficult to understand. Could you clarify your point?

From my understanding, the original playback speed issue at https://github.com/jellyfin/jellyfin-androidtv/issues/1517#issuecomment-1075723020 was closed because it's related to upstream developers (ExoPlayer). If that's the case, wouldn't it be necessary to address the problem there? If not, I’d appreciate some guidance on the best course of action.

**Comment by FunctionDJ** [comment]
I was able to fix the issue for me (at least temporarily) by either changing the preferred player to LibVLC, changing the Audiooutput to Mixdown Stereo (from Direct) or disabling DTS Audio Bitstream.

I'm wording it so weird because i changed all three settings (but then immediately changed player back to ExoPlayer), then force stopped the app, opened it, and the issue was fixed, but then afterwards i tried resetting the other two settings and the issue still doesn't come up even though all settings are exactly as before as far as i can tell.

So i'm not able to pinpoint exactly which setting fixed the issue for me, but maybe this info helps someone else.

**Comment by w1nwin** [comment]
All I did was change "Audio output" to "Downmix to stereo" and it worked again.

**Comment by xFastz** [comment]
I had a similar issue - did some tests with the following results:

Setup: Chromecast 4K plugged into Samsung Q990D soundbar, connected with eARC to LG G4 TV.

I realised the key issue with changing playback speed is due to AUDIO.
For files that required bitstream (dolby digital, DD+, Dolby atmos) etc, the playback speed is only 1x, regardless of what is being selected.
However, for other files that have audio tracks with PCM (shows up as either Stereo PCM or Multi Channel PCM on my soundbar), those files could have their playback speed changed.
For bitstream videos, I could change audio from "Direct" to "Mixdown Stereo" (as mentioned in above posts), and this would be transcoded and received by my soundbar as PCM, only then could I change playback speed, but losing much detail in the audio track.
Similarly, if I turn off "Bitstream" in the advanced audio settings, the audio will transcode and I will be able to speed up.

Have not managed to successfully get a bitstream audio sped up so far.

Tested all sorts of variations including plugging Chromecast direct to TV, setting passthrough etc, and the above were my conclusions. Please let me know if any other info would be helpful.

**Comment by FunctionDJ** [comment]
From further experience i think this is also related to the current video type / codec. I generally only have this problem when some DoVi profiles are active (without transcoding!).

On non-DoVi (and generally h264) content, or when transcoding, i don't think i've ever had this issue.

**Comment by player8472** [comment]
This definitely still happens (or doesn't happen...).

Speed setting mostly works for movies, but never for TV shows.
Using GoogleTV app on Sony Bravia

**Comment by ahmadmucom** [comment]
Still the same issue, I had to transcode my entire library to include an "AAC" encoded track if I wanted to increase the playback speed. Increasing the playback speed doesn't work with Dolby Atmos or DTS codecs

**Comment by MaienM** [comment]
This issue is still present.

**Comment by ahmadmucom** [comment]
> This issue is still present.

Yeah unfortunately and it's ruining the experience for me, not sure why no more people are complaining about it.

**Comment by FunctionDJ** [comment]
And also unfortuantely [Wolphin has a similar issue](https://github.com/damontecres/Wholphin/issues/164)

**Comment by ahmadmucom** [comment]
I implemented a fix for playback speed control with EAC3/DTS/Dolby Atmos audio codecs. Since I'm not an Android developer, I worked with AI assistance to develop this solution.

### Technical Approach

The issue is that Android TV hardware audio decoders (used for passthrough/offload) don't support playback speed changes for compressed formats like EAC3, DTS, and Dolby Atmos. The solution dynamically switches between two decoder strategies based on playback speed:

1. **At 1.0x speed (normal playback):**
   - Uses platform decoders with audio offload enabled
   - Full Dolby Atmos/TrueHD passthrough works
   - Maximum audio quality preserved

2. **At any other speed (>1.0x or <1.0x):**
   - Recreates the ExoPlayer instance with FFmpeg extension renderer mode preferred
   - Forces software decoding to PCM (uncompressed audio)
   - Playback speed changes work correctly
   - Audio is decoded to PCM (no spatial metadata like Atmos, but lossless decode quality)

### Implementation Details

The fix modifies:
- `app/src/main/java/org/jellyfin/androidtv/ui/playback/VideoManager.java` - Legacy playback system
- `playback/media3/exoplayer/src/main/kotlin/ExoPlayerBackend.kt` - New playback system

When speed changes between 1.0x and non-1.0x, the player is recreated with the appropriate renderer configuration. This causes a brief pause (~1-2 seconds) during the transition, which is acceptable for this use case.

### Known Limitations

- Brief pause when switching between 1.0x and other speeds (player recreation)
- Dolby Atmos/spatial audio is lost when speed ≠ 1.0x (decoded to PCM)
- Affects all audio formats, not just EAC3/DTS (for consistency)

### Why This Solution

Playback speed control is essential for my workflow, and I prioritize it over maintaining Atmos during speed changes. I was previously transcoding my library to include AAC tracks, which was time-consuming and reduced quality. This solution provides:
- Full quality (Atmos) at normal speed
- Working speed control when needed
- No need to pre-transcode media files

### Future Maintenance

I plan to maintain this as a soft fork, rebasing every few months to stay current with upstream changes. I'll test after each rebase to ensure compatibility.

### Build Artifact

A debug APK built with this fix is available here:
https://github.com/AhmadMuj/jellyfin-androidtv/actions/runs/19678362131

### Potential Integration

I'm not opening a PR at this time to avoid wasting maintainer time, since this approach has trade-offs. If the team is interested, I could contribute a settings flag to enable/disable this behavior, allowing users to choose between Atmos quality and speed control flexibility.

**Comment by Mysticodex** [comment]
Happening for me. Is this fixed for anyone yet?
